audit/base/base.go

```diff
@@ -0,0 +1,556 @@
+	}
+
+	cli := &Client{
+		settings:          Settings{}.defaults(),
+		log:               slog.Default(),
+		heartbeatInterval: 30 * time.Minute,
+		done:              make(chan struct{}),
+	}
+	cli.conn.Store(&c)
+
+	defer func() {
+		if err != nil {
+			c.CloseSend(ctxBack)
+		}
+	}()
+
+	for _, o := range options {
+		if err := o(cli); err != nil {
+			return nil, err
+		}
+	}
+
+	kver, err := kernelVer()
+	if err != nil {
+		return nil, fmt.Errorf("could not determine kernel version on platform: %w", err)
+	}
+
+	cli.heartbeat = msgs.Msg{
+		Type: msgs.Heartbeat,
+		Heartbeat: msgs.HeartbeatMsg{
+			AuditVersion: version.Semantic,
+			OsVersion:    kver,
+			Language:     runtime.Version(),
+			Destination:  c.Type().String(),
+		},
+	}
+	// This handles a special case where the version of the language is not goX.X.X, but instead
+	// the git hash of the compiler. This happens on custom builds of the go compiler.
+	if !strings.HasPrefix(cli.heartbeat.Heartbeat.Language, "go") {
+		cli.heartbeat.Heartbeat.Language = "go" + cli.heartbeat.Heartbeat.Language
+	}
+
+	cli.sendCh = make(chan streamData[SendMsg], cli.settings.QueueSize)
+	cli.stopSend = make(chan chan struct{}, 1)
+
+	go cli.sender()
+
+	return cli, nil
+}
+
+// Conn returns the current connection to the audit server. The one inside
+// Client can be changed out from under you. This is provided for testing purposes.
+// This is not supported for users.
+func (c *Client) Conn() conn.Audit {
+	ptr := c.conn.Load()
+	if ptr == nil {
+		return nil
+	}
+	return *ptr
+}
+
+// Logger returns the logger for the audit client.
+func (c *Client) Logger() *slog.Logger {
+	if c == nil || c.log == nil {
+		return slog.Default()
+	}
+	return c.log
+}
+
+/*
+Send sends an audit record to the audit server. Send is asynchronous and thread safe.
+
+Send is designed around speed. It will return immediately if the queue is full or your message
+doesn't validate. If the queue is full, Send() will return an error of type Error set to ErrQueueFull.
+You can detect this using IsQueueFull(). It is up to the caller to handle this error and resend
+the message.
+
+If the queue is not full and the message validates, Send() will return nil. This means that the message was
+queued to be sent to the audit server, not that the message was successfully sent. Send() is an asyncronous
+method.
+
+Any other errors are due to an unrecoverable error (such as a connection problem). If you get an error
+from Send() that IsUnrecoverable() returns true to, you should not use the client anymore. The messages
+that occur due to unrecoverable errors are put back into the  queue until that queue becomes full.
+Whenever a ErrQueueFull message occurs, the message being sent is dropped and you must deal with the message
+yourself.
+
+Send honors Context cancellation. If the context is canceled, the send will get cancelled. A Context timeout
```

Comment:
Wait does the context apply to putting the message into the queue or actually sending the message?

Since `Send` is async and doesn't actually send the message, it seems like the context cant apply to actually sending the message?

Oh wait no I see we actually capture the context and pass it into the queue.

This feels like an anti-feature to me (I think you agree). Is there an actual use-case where this is useful?

Reply:
Originally this was queuing AND sending the message. But since we changed:
`func (c *Conn) Write(ctx context.Context, msg msgs.Msg) error {`

I have also removed ctx.Done() from below. We just aren't going to honor this in an async client.